SC_redesign/Crop_and_Soil/crop/phosphorus_uptake.py

```diff
@@ -0,0 +1,316 @@
+            half_mature_fractional_phosphorus: fraction of phosphorus in the biomass at 50% maturity
+            fraction_potential_heat_units_full_maturity: fractional potential heat unit at 100% maturity
+            near_mature_fractional_phosphorus: fraction of phosphorus in the biomass at near 100% maturity
+
+        Returns:
+            the second shape coefficient
+    """
+
+    first_term = calc_logarithmic_term_of_coefficient(fraction_potential_heat_units_half_maturity,half_mature_fractional_phosphorus)
+    second_term = calc_logarithmic_term_of_coefficient(fraction_potential_heat_units_full_maturity,near_mature_fractional_phosphorus)
+
+    third_term = fraction_potential_heat_units_full_maturity - fraction_potential_heat_units_half_maturity
+    return (first_term - second_term) / third_term
+
+
+def calc_first_shape_coefficient(second_shape_coefficient: float,fraction_potential_heat_units_half_maturity: float,
+                                half_mature_fractional_phosphorus: float) -> float:
+    """Description:
+            Calculates the first shape coefficient
+
+        Args:
+            second_shape_coeffiient: 
+            fraction_potential_heat_units_half_mature: fractional potential heat units at 50% maturity
+            half_mature_fractional_phosphorus: fraction of phosphorus in the biomass at 50% maturity
+
+        Returns:
+            the first shape coefficient
+
+    """
+    
+    first_term = calc_logarithmic_term_of_coefficient(fraction_potential_heat_units_half_maturity, half_mature_fractional_phosphorus)
+
+    return first_term + second_shape_coefficient * fraction_potential_heat_units_half_maturity
+
+
+def calculate_fractional_biomass_phosphorus(first_shape_coefficient: float,second_shape_coefficient: float,
+                    previous_actual_biomass: float, emergent_fractional_biomass: float, mature_fractional_biomass: float,
+                    previous_fraction_potential_heat_units: float) -> None:
+
+    if previous_actual_biomass == 0:
+        return 0
+    first_term = emergent_fractional_biomass - mature_fractional_biomass
+    exponent_part = exp(first_shape_coefficient - second_shape_coefficient * previous_fraction_potential_heat_units)
+    second_term = 1 - (previous_fraction_potential_heat_units / (previous_fraction_potential_heat_units + exponent_part))
+
+    return first_term * second_term + mature_fractional_biomass
+
+
+
+#TODO: option1/2 need better names
+#calc_P_up
+def calc_potential_phosphorus_uptake(optimal_phosphorus_biomass: float, actual_phosphorus_biomass: float,
+                                mature_fractional_phosphorus: float ,potential_phosphorus_biomass_increase: float) -> float:
+    """
+    Description:
+        Calculates potential phosphorus uptake
+
+    Args:
+        optimal_phosphorus_biomass: optimal mass of phosphorus stored in plant material (kg P/ha)
+        actual_phosphorus_biomass: actual mass of phosphorus stored in plant material (kg P/ha)
```

Comment:
Here and elsewhere, remove the compound names from units. Phosphorus is measured in kg per hectare, not kg of phosphorus per hectare.

```suggestion
        optimal_phosphorus_biomass: optimal mass of phosphorus stored in plant material (kg/ha)
        actual_phosphorus_biomass: actual mass of phosphorus stored in plant material (kg/ha)
```